neuroanalysis/event_detection.py

```diff
@@ -95,45 +95,175 @@ def zero_crossing_events(data, min_length=3, min_peak=0.0, min_sum=0.0, noise_th
+        off_inds = list(np.argwhere(diff==-1)[:,0]) #where crosses from above to below threshold. Note taking index before this happens
+
+        if debug:
+            mpl.figure()
+            mpl.plot(data1, '.-') 
+            for on in on_inds:
+                mpl.axvline(x=on, color='g', linestyle='--')
+            for off in off_inds:
+                mpl.axvline(x=off, color='r', linestyle='--')
+            mpl.plot(threshold, color='y', linestyle='--')
+            mpl.plot(-threshold, color='y', linestyle='--')
+            mpl.show(block=False)
+
+
+
+        # sometimes an event happens at the beginning of the pulse window and the trace hasn't 
+        # been able to drop below threshold because it hasn't recovered from the artifact.
+        if len(off_inds) > 0:  #if there are some off indicies
+            if len(on_inds) > 0: #and there are also on indicies
+                if on_inds[0] > off_inds[0]: #check if off happens before on
+                    on_inds, off_inds = deal_unbalanced_initial_off(omit_ends, on_inds, off_inds)
+            else: #there are no on indicies
+                on_inds, off_inds = deal_unbalanced_initial_off(omit_ends, on_inds, off_inds)
+
+        if len(on_inds) > 0:  #if there are some on indicies
+            if len(off_inds) > 0: #and there are also off indicies
+                if on_inds[-1] > off_inds[-1]: #check if off happens before on
+                    on_inds, off_inds = deal_unbalanced_termination_on(omit_ends, on_inds, off_inds, len(data1))
+            else: #there are no off indicies
+                on_inds, off_inds = deal_unbalanced_termination_on(omit_ends, on_inds, off_inds, len(data1))
+
+        # this is insufficient because it ignores when there are just off or just on
+        # not sure why this is here if haven't decided whether or not to omit ends
+        # if len(on_inds) == 0 or len(off_inds) == 0:
+        #     continue    
+        # ## if there are unequal number of crossing from one direction, either remove the ends or add the appropriate initial or end index (which will be the beginning or end of the waveform)
+        # if off_inds[0] < on_inds[0]:  
+        #     if omit_ends:
+        #         off_inds = off_inds[1:]
+        #         if len(off_inds) == 0:
+        #             continue
+        #     else:
+        #         on_inds.insert(0, 0)
+        # if off_inds[-1] < on_inds[-1]:
+        #     if omit_ends:
+        #         on_inds = on_inds[:-1]
+        #     else:
+        #         off_inds.append(len(diff))
+        
+        # Add both events above +threshold and those below -threshold to a list
         for i in range(len(on_inds)):
-            if on_inds[i] == off_inds[i]:
+            
+            # remove any point where an on off is seperated by less than 2 indicies
+            if (on_inds[i] + 1) == off_inds[i]: 
```

Comment:
Why? Again, this is a generic threshold-crossing function; we shouldn't make assumptions about the use case here.

Reply:
This was useful in the downstream stream code, I can pull it out and put in the other function.